presto-main/src/main/java/com/facebook/presto/geospatial/H3HexFunctions.java

```diff
@@ -0,0 +1,367 @@
+
+public final class H3HexFunctions
+{
+    private static final H3CoreV3 h3;
+
+    private H3HexFunctions() {}
+
+    static {
+        try {
+            h3 = H3CoreV3.newInstance();
+        }
+        catch (IOException ex) {
+            throw new RuntimeException("Failed to load H3Core", ex);
+        }
+    }
+
+    private static Block convertHexagonAddrArrayToBlock(List<String> hexAddrs)
+    {
+        BlockBuilder block = VARCHAR.createBlockBuilder(null, hexAddrs.size());
+        for (String hexAddr : hexAddrs) {
+            VARCHAR.writeString(block, hexAddr);
+        }
+        return block.build();
+    }
+
+    private static List<String> convertBlockToHexagonAddrs(Block hexAddrBlock)
+    {
+        List<String> hexAddrsStr = new ArrayList<>();
+        for (int i = 0; i < hexAddrBlock.getPositionCount(); i++) {
+            Slice hexAddr = VARCHAR.getSlice(hexAddrBlock, i);
+            if (hexAddr.length() == 0) {
+                throw new PrestoException(INVALID_FUNCTION_ARGUMENT, "Input array contains an empty value.");
+            }
+            String hexAddrStr = unwrapAddress(hexAddr);
+            hexAddrsStr.add(hexAddrStr);
+        }
+        return hexAddrsStr;
+    }
+
+    private static AreaUnit getAreaUnit(String unitStr)
+            throws PrestoException
+    {
+        if (unitStr.equals("km2")) {
+            return AreaUnit.km2;
+        }
+        else if (unitStr.equals("m2")) {
+            return AreaUnit.m2;
+        }
+        else {
+            throw new PrestoException(INVALID_FUNCTION_ARGUMENT, "Area unit must be 'm2' or 'km2'.");
+        }
+    }
+
+    private static String unwrapAddress(@SqlType(StandardTypes.VARCHAR) Slice hexAddr)
+            throws PrestoException
+    {
+        String hexAddress = hexAddr.toStringUtf8();
+        if (!h3.h3IsValid(hexAddress)) {
+            throw new PrestoException(INVALID_FUNCTION_ARGUMENT, "Input is not a valid h3 address (" + hexAddress + ").");
+        }
+        return hexAddress;
+    }
+
+    private static LengthUnit getLengthUnit(String unitStr)
+            throws PrestoException
+    {
+        if (unitStr.equals("km")) {
+            return LengthUnit.km;
+        }
+        else if (unitStr.equals("m")) {
+            return LengthUnit.m;
+        }
+        else {
+            throw new PrestoException(INVALID_FUNCTION_ARGUMENT, "Length unit must be 'm' or 'km'.");
+        }
+    }
+
+    @Description("Returns hexagon address calculated from H3 library.\\nlat(double): latitude of the coordinate\\nlng(double): longitude of the coordinate\\nres(int): resolution of the address, between 0 and 15 inclusive")
+    @ScalarFunction("get_hexagon_addr")
+    @SqlType(StandardTypes.VARCHAR)
+    public static Slice getHexagonAddr(@SqlType(StandardTypes.DOUBLE) double lat,
+            @SqlType(StandardTypes.DOUBLE) double lng,
+            @SqlType(StandardTypes.BIGINT) long res)
```

Comment:
seems like the range of a resolution is between 0 and 15, how about using a smaller data type?